tests/components/risco/util.py

```diff
@@ -43,15 +58,36 @@ async def setup_risco(hass, events=[], options={}):
     return config_entry
 
 
+async def setup_risco_local(hass, options={}):
```

Comment:
Don't set a default arguments to a mutable object. That's error prone.

This function can be made a pytest fixture instead that depends on another fixture for the options. We can override fixtures per test via `pytest.mark.parametrize`. Use that to set custom options.

Reply:
Thanks.
How would you handle situations where we need to make assertions before setup? (e.g. https://github.com/home-assistant/core/pull/75874/files#diff-2a7abdb0037958212636f0163d689d80071e1a19f4946380cf12b81b07231534R65)

Also is there an example of a fixture dependent on another fixture? Not even sure what to search for.

Reply:
A fixture depends on another fixture if we pass the name of the fixture to depend on as an argument to the fixture that needs the other fixture. This is just standard pytest behavior. Please read more about that in the pytest docs.

Some cases may need the fixture to return or yield a function that can be used in the test as needed.

We often don't assert the lack of states before setting up an integration, though. Unless I'm doing some modifications to the entity or device state before setting up the integration, I'd skip asserting that there are no states at that point. An example where it could be interesting to assert the entity state before setting up the integration is if I'm enabling an entity that is normally disabled before setting up the integration.

I recommend making a separate fixture that returns a mock config entry. Then it's easy to set up the integration using that config entry, inside the test for the cases where we want more control.

Reply:
Sorry, should have made my question clearer. I know how to pass fixtures by name, and I know how to pass them dynamically via `pytest.lazy_fixture` (for which I couldn't find any reference in the HA codebase, so not sure whether it's frowned upon, or what are the best practices). What I don't know is how to *dynamically* pass a fixture to another one.

I removed the no-state assertions, but there are still other places where code is executed / functions are patched before calling `setup_*`.

In any case, I removed the mutable default arguments through `*args` and `**kwargs` for now, hope that it's satisfactory.

Reply:
Please explain the problem with a specific example.

Reply:
For example, [here](https://github.com/home-assistant/core/pull/75874/files#diff-79db24431d93f5c3a4de758b472607e4d6709b8b4b4037767cfeb7443f7a5b37R170) we have:
```python
hass.config.set_time_zone("UTC")
with patch(
        "homeassistant.components.risco.Store.async_save",
    ) as save_mock:
    await setup_risco_cloud(hass, *TEST_EVENTS)
...
```
We have 3 things as a dependency to `setup_risco_cloud`:
1. Setting the time zone
2. Patching `async_save`
3. The test events

Reply:
The order of a fixture arguments in the test signature will determine the order of fixture setup, unless the fixtures themselves have further dependencies that control the order.

Does that answer your question?

Reply:
It answers 1 & 2. As for 3, I guess that the latest call to patch is the one that determines the behavior? So if I patch from within `setup_risco_cloud` to return an empty list as events, and then have another fixture in a later argument that patches to return the test events it could work.
I'll try it later

Reply:
I'd extract the test events to a separate fixture. Then those can be overridden as needed per test.

Reply:
Patching the events with an empty list is needed everywhere `setup_risco_cloud` is used (with the exception of cases where we're patching with actual data), so it means adding that separate fixture to almost every test. Would the patch over patch approach not work?

Reply:
I'd make one fixture for `setup_risco_cloud` that patches and sets up the config entry and another fixture that returns the return value for the patch of the events. The cloud setup fixture can depend on the events fixture. The events fixture can return an empty list by default.

Reply:
OK, so that's exactly the question I previously had. Let's say we have a fixture called `events`, and we use the following signatures:
```python
@fixture
def events(*events):

@fixture
async def setup_risco_cloud(hass, events):
```

If I want to have a test with non-default events:
```python
async def test_cloud_setup(hass, two_zone_cloud, setup_risco_cloud):
```

How do I customize the events, given that the dependency on `events` is not in this signature, but rather in `setup_risco_cloud`?

Reply:
Use `pytest.mark.parametrize`. The fixture doesn't need to be present in the signature of the current test to override it by parametrization.

Reply:
Here's one example where we don't even define the `platforms` fixture function. We just override it per test.

https://github.com/home-assistant/core/blob/58b9785485af4b49097707edb7fbcc00c72a3df0/tests/components/evil_genius_labs/conftest.py#L39-L55

https://github.com/home-assistant/core/blob/58b9785485af4b49097707edb7fbcc00c72a3df0/tests/components/evil_genius_labs/test_light.py#L15-L16

Reply:
Perfect - just what I was looking for